GitUI/UserControls/RevisionGrid/RevisionGridControl.cs

```diff
@@ -490,94 +455,21 @@ public void DisableContextMenu()
-                    {
-                        revListArgs += "--grep=\"" + revisionFilter.Text + "\" ";
-                    }
-                    else
-                    {
-                        inMemMessageFilter = revisionFilter.Text;
-                    }
-                }
-
-                if (revisionFilter.FilterByCommitter && !string.IsNullOrWhiteSpace(revisionFilter.Text))
-                {
-                    if (cmdLineSafe)
-                    {
-                        revListArgs += "--committer=\"" + revisionFilter.Text + "\" ";
-                    }
-                    else
-                    {
-                        inMemCommitterFilter = revisionFilter.Text;
-                    }
-                }
-
-                if (revisionFilter.FilterByAuthor && !string.IsNullOrWhiteSpace(revisionFilter.Text))
-                {
-                    if (cmdLineSafe)
-                    {
-                        revListArgs += "--author=\"" + revisionFilter.Text + "\" ";
-                    }
-                    else
-                    {
-                        inMemAuthorFilter = revisionFilter.Text;
-                    }
-                }
-
-                if (revisionFilter.FilterByDiffContent)
-                {
-                    if (cmdLineSafe)
-                    {
-                        revListArgs += "-G" + revisionFilter.Text.Quote();
-                    }
-                    else
-                    {
-                        throw new InvalidOperationException(string.Format(_invalidDiffContainsFilter.Text, revisionFilter.Text));
-                    }
-                }
-            }
-        }
-
-        public void SetAndApplyBranchFilter(string filter, bool requireRefresh)
+        public void SetAndApplyBranchFilter(string filter)
         {
-            AppSettings.BranchFilterEnabled = !string.IsNullOrWhiteSpace(filter);
+            // TODO: clean up and move all internals to FilterInfo
 
             // ShowCurrentBranchOnly depends on BranchFilterEnabled, and to show the current branch
             // both flags must be set simultaneously (check SetShowBranches implementation).
             // And since we set a filter - we can't be showing the current branch.
-            AppSettings.ShowCurrentBranchOnly = false;
+            _filterInfo.ShowCurrentBranchOnly = false;
 
-            requireRefresh = requireRefresh && _formRevisionFilter.SetBranchFilter(filter);
+            string newFilter = filter?.Trim() ?? string.Empty;
+            _filterInfo.ByBranchFilter = !string.IsNullOrWhiteSpace(newFilter);
+            _filterInfo.BranchFilter = newFilter;
 
-            if (requireRefresh)
-            {
-                ForceRefreshRevisions();
-            }
-            else
-            {
-                SetShowBranches();
-            }
-
-            OnFilterChanged();
+            // TODO: there's an opportunity to reduce refreshes by tracking the current state
```

Comment:
How did you see that `requireRefresh` is not required?
I tried to backtrack this briefly and it could be OK to remove.

Reply:
It was always set to `true`, except in some odd scenario when we reset the filter before switching the repo (or something like that). I have reworked the whole structure and the flow of filters, and this flag is superficial. We always need to refresh the grid whenever a filter changes.

I added the TODO note here because we can optimise the flow, and possibly skip a refresh if we detect that the branch filter value hasn't changed. But this is a function of the revision grid, and not the caller.